### PR TITLE
Make delete async

### PR DESCRIPTION
**Pull request by frobware**
Testing is minimal and would need further rework which can be done as
a follow-up PR.

### CODE REVIEW

**Comment by frobware** [comment]
I don't know whether we care for updating the status during delete, just added it there for consistency (and some visual debugging).

File: pkg/cloud/gcp/actuators/machine/reconciler.go [hunk 1]
```diff
@@ -273,21 +273,21 @@ func (r *Reconciler) exists() (bool, error) {
 
 // Returns true if machine exists.
 func (r *Reconciler) delete() error {
-	exists, err := r.exists()
+	instance, err := r.computeService.InstancesGet(r.projectID, r.providerSpec.Zone, r.machine.Name)
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: Reverted to calling `exists()`; i only dropped / changed it to prevent another call to `InstancesGet`() - but I also dropped status updates. YMMV.

File: pkg/cloud/gcp/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -273,21 +273,21 @@ func (r *Reconciler) exists() (bool, error) {
...
-	if !exists {
-		klog.Infof("Machine %v not found during delete, skipping", r.machine.Name)
+	if instance == nil {
+		klog.Infof("Machine %q not found during delete, skipping", r.machine.Name)
 		return nil
 	}
-	zone := r.providerSpec.Zone
-	operation, err := r.computeService.InstancesDelete(r.projectID, zone, r.machine.Name)
-	if err != nil {
+	if _, err = r.computeService.InstancesDelete(r.projectID, r.providerSpec.Zone, r.machine.Name); err != nil {
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: Taken both suggestions.

File: pkg/cloud/gcp/actuators/machine/reconciler.go [hunk 3]
```diff
@@ -273,21 +273,21 @@ func (r *Reconciler) exists() (bool, error) {
...
 		return fmt.Errorf("failed to delete instance via compute service: %v", err)
 	}
-	if op, err := r.waitUntilOperationCompleted(zone, operation.Name); err != nil {
-		return fmt.Errorf("failed to wait for delete operation via compute service. Operation status: %v. Error: %v", op, err)
+	r.providerStatus.InstanceState = &instance.Status
+	if instance.Status != "TERMINATED" {
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: Done - Inverted the status check in `exists()`.

File: pkg/cloud/gcp/actuators/machine/reconciler.go [hunk 4]
```diff
@@ -256,12 +256,12 @@ func (r *Reconciler) exists() (bool, error) {
 	instance, err := r.computeService.InstancesGet(r.projectID, zone, r.machine.Name)
 	if err == nil {
 		switch instance.Status {
-		case "PROVISIONING", "REPAIRING", "RUNNING", "STAGING":
-			klog.Infof("Machine %q already exists", r.machine.Name)
-			return true, nil
-		default:
+		case "TERMINATED":
 			klog.Infof("Machine %q is considered as non existent as its status is %q", instance.Status)
```
Comment (enxebre) [inline pkg/cloud/gcp/actuators/machine/reconciler.go]: this belong to different commit

Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: This would break the delete behaviour. We want to keep re queuing until TERMINATED.

Comment (enxebre) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: yeh, I'd do first commit: "change exist to only false on terminated", second commit: "make delete async", not a blocker to me though.
Also we can drop the polling function now

Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: > yeh, I'd do first commit: "change exist to only false on terminated",

https://github.com/openshift/cluster-api-provider-gcp/pull/25

**Comment by frobware** [comment]
/hold

Waiting on https://github.com/openshift/cluster-api-provider-gcp/pull/25

File: pkg/cloud/gcp/actuators/machine/reconciler.go (hunk 4 above)
Comment (frobware) [reply pkg/cloud/gcp/actuators/machine/reconciler.go]: > Also we can drop the polling function now

Done in 82a3f13ad1ae3c4c191e54957a226170c5bc63e2

**Comment by frobware** [comment]
/hold

Manual / sanity testing.

**Comment by frobware** [comment]
Runtime behaviour: 

```console
I0531 13:37:34.187256    1703 controller.go:193] Reconciling machine "amcdermo-test-gcp-1" triggers delete
I0531 13:37:34.187268    1703 actuator.go:89] Deleting machine amcdermo-test-gcp-1
I0531 13:37:34.712075    1703 reconciler.go:237] Machine "amcdermo-test-gcp-1" already exists
I0531 13:37:35.256189    1703 reconciler.go:261] machine "amcdermo-test-gcp-1" status is exists, requeuing...
E0531 13:37:35.256266    1703 controller.go:207] Failed to delete machine "amcdermo-test-gcp-1": requeue in: 20s
I0531 13:37:35.256297    1703 controller.go:346] Actuator returned requeue-after error: requeue in: 20s
I0531 13:37:55.256673    1703 controller.go:129] Reconciling Machine "amcdermo-test-gcp-1"
I0531 13:37:55.256724    1703 controller.go:292] Machine "amcdermo-test-gcp-1" in namespace "test-gcp" doesn't specify "cluster.k8s.io/cluster-name" label, assuming nil cluster
I0531 13:37:55.256763    1703 controller.go:193] Reconciling machine "amcdermo-test-gcp-1" triggers delete
I0531 13:37:55.256788    1703 actuator.go:89] Deleting machine amcdermo-test-gcp-1
I0531 13:37:55.822772    1703 reconciler.go:237] Machine "amcdermo-test-gcp-1" already exists
I0531 13:37:56.232196    1703 reconciler.go:261] machine "amcdermo-test-gcp-1" status is exists, requeuing...
E0531 13:37:56.232281    1703 controller.go:207] Failed to delete machine "amcdermo-test-gcp-1": requeue in: 20s
I0531 13:37:56.232315    1703 controller.go:346] Actuator returned requeue-after error: requeue in: 20s
I0531 13:38:16.232527    1703 controller.go:129] Reconciling Machine "amcdermo-test-gcp-1"
I0531 13:38:16.232567    1703 controller.go:292] Machine "amcdermo-test-gcp-1" in namespace "test-gcp" doesn't specify "cluster.k8s.io/cluster-name" label, assuming nil cluster
I0531 13:38:16.232579    1703 controller.go:193] Reconciling machine "amcdermo-test-gcp-1" triggers delete
I0531 13:38:16.232597    1703 actuator.go:89] Deleting machine amcdermo-test-gcp-1
I0531 13:38:16.780866    1703 reconciler.go:234] Machine "amcdermo-test-gcp-1" is considered as non existent as its status is "TERMINATED"
I0531 13:38:16.780913    1703 reconciler.go:255] Machine amcdermo-test-gcp-1 not found during delete, skipping
I0531 13:38:16.894092    1703 controller.go:226] Machine "amcdermo-test-gcp-1" deletion successful
```
